src/gui/fileactivitymodel.h

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (C) by Felix Weilbach <felix.weilbach@nextcloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+#pragma once
+
+#include <unordered_map>
+
+#include <QObject>
+#include <QAbstractItemModel>
+#include <QMetaType>
+#include <QIcon>
+#include <QTimer>
+#include <QDateTime>
+
+#include "activityjob.h"
+#include "tray/ActivityData.h"
+
+namespace OCC {
+
+class FileActivity
+{
+public:
+    enum class Type {
+        Unknown,
+        Created,
+        Shared,
+        Changed,
+    };
+
+    FileActivity();
+
+    FileActivity(int id, const QString &message, const QDateTime &timestamp, Type icon);
+
+    int id() const;
+
+    QString message() const;
+
+    void setMessage(const QString &message);
+
+    QDateTime timestamp() const;
+
+    void setTimestamp(const QDateTime &dateTime);
+
+    Type type() const;
+
+    void setType(Type type);
+
+private:
+    int _id;
+    QString _message;
+    QDateTime _timestamp;
+    Type _type;
+};
+
+class FileActivityListModel : public QAbstractListModel
+{
+public:
+    explicit FileActivityListModel(QObject *parent = nullptr);
+
+    void addFileActivity(const FileActivity &fileActivity);
+
+    void addFileActivities(const std::vector<FileActivity> &fileActivities);
+
+    int rowCount(const QModelIndex &parent = QModelIndex()) const override;
+
+    QVariant data(const QModelIndex &index, int role = Qt::DisplayRole) const override;
+
+    QVariant headerData(int section, Qt::Orientation orientation, int role = Qt::DisplayRole) const override;
+
+    static std::shared_ptr<QPixmap> pixmapForActivityType(FileActivity::Type type, int size);
+
+private:
+    static std::shared_ptr<QPixmap> getIconPixmap(const QString &iconName, int size);
+
+    std::unordered_map<int, std::tuple<int, FileActivity *>> _fileActivityMap;
```

Comment:
@FlexW Why use `std::tuple` instead of `std::pair`?

Reply:
@FlexW If I was Kevin, I'd even as to use a struct instead :)

Reply:
I can recall that discussion, and I even remember that I said I would use a struct too :D The problem here was that I could not come up with a name for that struct

Reply:
@FlexW Ok. Still, tuple seems like overkill for 2 values IMHO.